tree/dataframe/src/RDFDisplayer.cxx

```diff
@@ -0,0 +1,233 @@
+{
+   size_t totalWidth = 0;
+
+   auto size = fWidths.size();
+   for (size_t i = 0; i < size; ++i) {
+      totalWidth += fWidths[i];
+      if (totalWidth > fgMaxWidth) {
+         return size - i;
+      }
+   }
+
+   return 0;
+}
+
+void RDisplayer::Print() const
+{
+   auto columnsToPrint =
+      fNColumns - GetNColumnsToShorten(); // Get the number of columns that fit in the characters limit
+   std::vector<bool> hasPrintedNext(fNColumns,
+                                    false); // Keeps track if the collection as already been shortened, allowing to skip
+                                            // all elements until the next printable element.
+   auto nrRows = fTable.size();
+   for (size_t rowIndex = 0; rowIndex < nrRows; ++rowIndex) {
+      auto &row = fTable[rowIndex];
+
+      std::stringstream stringRow;
+      bool isRowEmpty = true; // It may happen during compacting that some rows are empty, this happens for example if
+                              // collections have different size. Thanks to this flag, these rows are just ignored.
+      for (size_t columnIndex = 0; columnIndex < columnsToPrint; ++columnIndex) {
+         const auto &element = row[columnIndex];
+         std::string printedElement = "";
+
+         if (element.HasToBeDotted()) {
+            printedElement = "...";
+         } else if (element.HasToBePrinted()) {
+            // Maybe the element is part of a collection that is being shortened, and so it was already printed.
+            if (!hasPrintedNext[columnIndex]) {
+               printedElement = element.GetRepresentation();
+            }
+            hasPrintedNext[columnIndex] =
+               false; // Encountered "next printable element", shortening can start again when needed.
+         } else {     // HasToBeIgnored
+            // Shortening is starting here. Print directly the last element, to have something like 1 ... 3, and don't
+            // print anything else.
+            if (!hasPrintedNext[columnIndex]) {
+               size_t i = rowIndex + 1; // Starting from the next row...
+               for (; !fTable[i][columnIndex].HasToBePrinted(); ++i) {
+                  // .. look for the first element that can be printed, it will be the last of the collection.
+               }
+               printedElement = fTable[i][columnIndex].GetRepresentation(); // Print the element
+               hasPrintedNext[columnIndex] = true; // And start ignoring anything else until the next collection.
+            }
+         }
+         if (!printedElement.empty()) {
+            // Found at least one element, so the row is not empty.
+            isRowEmpty = false;
+         }
+
+         stringRow << std::left << std::setw(fWidths[columnIndex]) << std::setfill(fgSeparator) << printedElement
+                   << " | ";
+      }
+      if (!isRowEmpty) {
+         std::cout << stringRow.str() << std::endl;
+      }
+   }
+}
+
+std::string RDisplayer::AsString() const
```

Comment:
Can't print depend on AsString?

Reply:
I thought a lot about this, and it really can't, because they do really different things.
What I did was making the Print code much more complex to avoid storing the data in two separate structures. In this way, Print() and AsString() both share the same data.